README.md
# Meta Snippet Linter

Read the titles and meta descriptions a build actually shipped, count them in one explicitly chosen
unit, and hold them against one explicit policy: per-locale length bounds, uniqueness within a
declared scope, and coverage of every page the configuration says exists — each finding carrying the
file, the element and the line it came from.

- **Repository:** [edilec/meta-snippet-linter](https://github.com/edilec/meta-snippet-linter)
- **Area:** SEO & Search
- **License:** MIT
- **Dependencies:** none. Node 22+ built-ins only.

```
npx meta-snippet-linter --config site/meta.config.json
```

## What it does

```
$ meta-snippet-linter --config examples/broken/meta.config.json
meta-snippet-linter: fail
  4 of 5 page(s) measured, 1 skipped, 2 duplicate group(s)
  9 error(s), 2 warning(s), 0 info
  [error] title-duplicate build/copy.html:5:5 /head/title
      Page "/copy" shares its title with 1 other page(s) in locale "en": /.
  ...
  note: Length bounds are house-style heuristics, not search-engine guarantees: ...
```

stdout carries the JSON report and nothing else, so it pipes straight into a parser. The summary
above is on stderr, and `--json` silences it.

## Configuration

```json
{
  "schemaVersion": "1",
  "counting": "graphemes",
  "duplicateScope": "locale",
  "defaults": {
    "title": { "min": 20, "max": 60 },
    "description": { "min": 70, "max": 155 }
  },
  "locales": {
    "en": {},
    "ja": { "title": { "min": 8, "max": 32 }, "description": { "min": 30, "max": 90 } }
  },
  "pages": [
    { "path": "/", "file": "build/index.html", "locale": "en" },
    { "path": "/ja/", "file": "build/ja-index.html", "locale": "ja" }
  ]
}
```

`counting` and `duplicateScope` are required and have no default: both change the answer, and a
report that guessed either would not be saying what it measured. A locale entry inherits `defaults`
and overrides only the bounds it names. Unknown keys anywhere are refused, not ignored.

Runnable examples live in `examples/clean` (passes, exit 0) and `examples/broken` (fails, exit 1).

## Counting, exactly

Counting is the subtle part, so the choice is explicit and documented rather than assumed. Every
snippet is decoded, whitespace-collapsed and **NFC-normalised** before it is counted and before it
is grouped, so two titles that render identically always count identically — and `text-not-nfc`
tells you when a page's source needed normalising.

| Unit | `⚠️` (U+26A0 U+FE0F) | `👩‍👩‍👧` | `Café` (decomposed) |
| --- | ---: | ---: | ---: |
| `graphemes` | 1 | 1 | 4 |
| `codePoints` | 2 | 5 | 4 |
| `utf16CodeUnits` | 2 | 8 | 4 |

`graphemes` uses the built-in `Intl.Segmenter`. Full details, including what NFC normalisation does
to your numbers, are in [docs/meta-snippet-rules.md](./docs/meta-snippet-rules.md).

## Exit codes

| Code | Meaning |
| ---: | --- |
| `0` | every declared page was measured and the policy was satisfied |
| `1` | the build failed the policy |
| `2` | a configuration error (stdout empty), or evidence missing, undecodable or bounded out (an `incomplete` report on stdout) |

`incomplete` outranks `fail`, which outranks `pass`. A run that could not read a page never exits 0,
and a run that measured nothing reports `nothing-checked` rather than coming back clean.

## Limits and non-goals

**The length bounds are house style, not search-engine guarantees.** No engine publishes a character
limit. Snippets are truncated by rendered pixel width in a layout nobody controls, on a device
nobody controls, and an engine may rewrite a title or description entirely from page content. A
title inside your bounds is a title that matches the policy you wrote down — nothing more. This tool
carries that sentence in its own report so it cannot be quoted as a ranking promise.

Things this tool **cannot** conclude:

- **Whether a snippet is good.** It measures length, uniqueness and presence. Relevance, accuracy,
  tone and keyword intent are outside what a file can tell it.
- **What a search engine will display, index or rank.** It never contacts one, and nothing in its
  output is evidence about one.
- **What a client-rendered page ends up showing.** It reads the shipped HTML. A title written by
  JavaScript after load, injected by a framework at runtime, or swapped by a client-side router is
  invisible to it.
- **That an unlisted page is fine.** Pages come only from the configuration; nothing is discovered
  by walking a directory. That is what makes runs reproducible, and it means coverage is exactly as
  complete as your page list.
- **That a page it could not read is fine.** An unreadable, oversized, non-UTF-8 or bounded-out page
  is `incomplete` and exit 2, never a pass. Only regular files are read: a directory or a named pipe
  standing where a page should be is refused rather than opened.
- **Anything about `og:description`, `twitter:description` or a `<base>`-relative rewrite.** Only
  `<title>` and `<meta name="description">` are read.
- **That the head is exactly what a browser would build.** The scanner reads forward through markup
  rather than constructing a DOM: it understands comments, doctypes, raw-text elements and quoted
  attributes, but it does not run scripts or repair broken nesting.
- **That an encoding other than UTF-8 was read correctly.** UTF-8 is the only encoding supported;
  anything else is decoded strictly, and a failure is reported rather than patched with replacement
  characters.

It reads files and nothing else: no network, no telemetry, no writes. A page path that leaves the
input root — by spelling, or through a symbolic link planted inside it — is refused before the file
is opened. Page text that does reach the report is quoted as a bounded, single-line excerpt with
control characters replaced, so nothing a build directory contains can repaint the terminal the
summary is printed in.

## Development

```
npm run check     # lint, test, run the clean example, and dry-run the package
npm test
npm run test:coverage
```

No dependencies, no install step. `docs/meta-snippet-rules.md` is the rule contract and the test
suite asserts the code against it in both directions.

## License

MIT. See [LICENSE](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project are recorded here. The format follows
[Keep a Changelog](https://keepachangelog.com/en/1.1.0/), and versions follow
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Renaming a `ruleId` is a breaking change and will be recorded here.

## [Unreleased]

### Fixed

- A config file that does not parse is no longer echoed back on stderr.
  `JSON.parse` embeds the input in one of its two error messages
  (`Unexpected token 'A', "AKIA…" is not valid JSON`), so a config short enough
  to be only a credential was reproduced in full by `The config is not valid
  JSON: …`. `parseFailureDetail` in `src/rules.mjs` keeps the position, line
  and column — which carry no input — and drops the quotation. Excerpting the
  message would not have helped: the quotation is at the front of the message
  and `excerpt` cuts from the back.

### Added

- Bounded head scanner for built HTML that reads `<title>`, `<meta name="description">`, the html
  `lang` attribute and a declared charset, with the source line and column of each.
- Explicit counting pipeline: character-reference decoding, ASCII whitespace collapsing, NFC
  normalisation, length bounding, then counting in `graphemes`, `codePoints` or `utf16CodeUnits`.
  `counting` is required and has no default.
- Per-locale length policy with `defaults` and locale overrides that merge rather than replace.
- Duplicate grouping within a required `duplicateScope` of `locale` or `site`.
- Locale scope and coverage rules: a page outside the declared locales is reported and excluded, and
  a declared page that was never built is `page-unreadable` and `incomplete`.
- Report contract v1 output on stdout with a human summary on stderr, a frozen `ruleId` to severity
  table, and deterministic ordering by file, pointer, rule, line, column and message.
- `bin/meta-snippet-linter.mjs` with `--help`, `--json`, `--root`, `--counting`,
  `--duplicate-scope` and a flag for every documented limit.
- `examples/clean` and `examples/broken`, and the rule contract in `docs/meta-snippet-rules.md`.

### Security

- Page paths are confined to the input root by real path, after symbolic links are resolved, so a
  link planted inside the build output cannot cause an out-of-root file to be read or quoted.
- Only regular files are read. A directory, a named pipe or a device node standing where a page
  should be is `page-unreadable` with the reason `not a regular file`, and is never opened.
- Report evidence is scrubbed of every C0, `DEL` and C1 control character and of the two Unicode
  line separators, bounded to 160 characters out and 8192 code units in, so page content cannot
  repaint the terminal the stderr summary is rendered in.

No release has been published.


test/cli.test.mjs
import assert from 'node:assert/strict'
import test from 'node:test'
import { execFile } from 'node:child_process'
import { readFile } from 'node:fs/promises'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'
import { promisify } from 'node:util'

const run = promisify(execFile)
const HERE = dirname(fileURLToPath(import.meta.url))
const ROOT = join(HERE, '..')
const CLI = join(ROOT, 'bin', 'meta-snippet-linter.mjs')
const fixture = (...parts) => join(HERE, 'fixtures', ...parts)
const example = (...parts) => join(ROOT, 'examples', ...parts)

async function cli(args) {
  try {
    const { stdout, stderr } = await run(process.execPath, [CLI, ...args])
    return { code: 0, stdout, stderr }
  } catch (error) {
    return { code: error.code, stdout: error.stdout, stderr: error.stderr }
  }
}

const rules = (report) => report.findings.map((finding) => finding.ruleId)

test('--help explains the tool, prints on stdout and exits 0', async () => {
  const result = await cli(['--help'])

  assert.equal(result.code, 0)
  assert.match(result.stdout, /^meta-snippet-linter 0\.1\.0/)
  assert.match(result.stdout, /not search-engine guarantees/)
  assert.match(result.stdout, /--counting UNIT/)
  assert.match(result.stdout, /--duplicate-scope S/)
  assert.match(result.stdout, /--max-text-length N/)
  assert.equal(result.stderr, '')
})

test('--version matches the published package version', async () => {
  const result = await cli(['--version'])
  const manifest = JSON.parse(await readFile(join(ROOT, 'package.json'), 'utf8'))

  assert.equal(result.code, 0)
  assert.equal(result.stdout, `${manifest.version}\n`)
})

test('the clean example exits 0 with a parseable report on stdout and a summary on stderr', async () => {
  const result = await cli(['--config', example('clean', 'meta.config.json')])
  const report = JSON.parse(result.stdout)

  assert.equal(result.code, 0)
  assert.equal(report.schemaVersion, '1')
  assert.equal(report.tool, 'meta-snippet-linter')
  assert.equal(report.status, 'pass')
  assert.equal(report.summary.checked, 3)
  assert.deepEqual(report.findings, [])
  assert.match(result.stderr, /^meta-snippet-linter: pass/)
  assert.match(result.stderr, /note: Length bounds are house-style heuristics/)
})

test('--json keeps stderr empty and leaves stdout parseable', async () => {
  const result = await cli(['--config', example('clean', 'meta.config.json'), '--json'])

  assert.equal(result.code, 0)
  assert.equal(result.stderr, '')
  assert.equal(JSON.parse(result.stdout).status, 'pass')
})

test('the broken example exits 1 and reports every page it measured', async () => {
  const result = await cli(['--config', example('broken', 'meta.config.json'), '--json'])
  const report = JSON.parse(result.stdout)

  assert.equal(result.code, 1)
  assert.equal(report.status, 'fail')
  assert.equal(report.summary.pages, 5)
  assert.equal(report.summary.checked, 4)
  assert.deepEqual(rules(report), [
    'description-missing',
    'title-too-long',
    'description-duplicate',
    'title-duplicate',
    'title-too-short',
    'locale-not-declared',
    'description-duplicate',
    'title-duplicate',
    'title-too-short',
    'title-repeated-element',
    'html-lang-mismatch',
  ])
})

test('an unreadable input exits 2 with an incomplete report on stdout', async () => {
  const result = await cli(['--config', fixture('incomplete', 'missing-page.config.json'), '--json'])
  const report = JSON.parse(result.stdout)

  assert.equal(result.code, 2)
  assert.equal(report.status, 'incomplete')
  assert.deepEqual(rules(report), ['page-unreadable'])
  assert.equal(report.findings[0].location.file, 'build/never-built.html')
})

test('a configuration error exits 2 with empty stdout and the message on stderr', async () => {
  const cases = [
    [['--config', fixture('invalid', 'unknown-key.config.json')], /Unknown config key "duplcateScope"/],
    [['--config', fixture('invalid', 'unknown-limit.config.json')], /Unknown limit "maxPage"/],
    [['--config', fixture('invalid', 'no-counting.config.json')], /counting must be one of/],
    [['--config', fixture('invalid', 'not-json.config.json')], /not valid JSON/],
  ]

  for (const [args, expected] of cases) {
    const result = await cli(args)
    assert.equal(result.code, 2, args.join(' '))
    assert.equal(result.stdout, '', args.join(' '))
    assert.match(result.stderr, expected)
  }
})

test('bad usage exits 2 with empty stdout and the help text', async () => {
  const cases = [
    [[], /--config is required/],
    [['--config'], /--config requires a value/],
    [['--config', example('clean', 'meta.config.json'), '--verbose'], /Unknown option "--verbose"/],
    [['--config', example('clean', 'meta.config.json'), '--max-pages', '0'], /--max-pages requires a positive integer/],
    [['--config', example('clean', 'meta.config.json'), '--max-pages', 'lots'], /--max-pages requires a positive integer/],
  ]

  for (const [args, expected] of cases) {
    const result = await cli(args)
    assert.equal(result.code, 2, args.join(' '))
    assert.equal(result.stdout, '', args.join(' '))
    assert.match(result.stderr, expected)
  }
})

test('stdout is byte-identical across runs', async () => {
  const first = await cli(['--config', example('broken', 'meta.config.json'), '--json'])
  const second = await cli(['--config', example('broken', 'meta.config.json'), '--json'])

  assert.equal(first.code, 1)
  assert.equal(first.stdout, second.stdout)
  assert.ok(first.stdout.length > 0)
})

test('every documented limit flag is wired through to the check', async () => {
  const cases = [
    [['--max-pages', '1'], 'page-limit-exceeded'],
    [['--max-html-bytes', '16'], 'page-too-large'],
    [['--max-head-length', '120'], 'head-scan-truncated'],
    [['--max-tags', '3'], 'head-scan-truncated'],
    [['--max-text-length', '8'], 'text-truncated'],
  ]

  for (const [flag, expected] of cases) {
    const result = await cli(['--config', example('clean', 'meta.config.json'), '--json', ...flag])
    const report = JSON.parse(result.stdout)
    assert.equal(result.code, 2, flag.join(' '))
    assert.equal(report.status, 'incomplete', flag.join(' '))
    assert.ok(rules(report).includes(expected), `${flag.join(' ')} should produce ${expected}`)
  }

  // The control: without any flag the same example is clean, so each case above
  // is caused by the flag and not by the fixture.
  const control = await cli(['--config', example('clean', 'meta.config.json'), '--json'])
  assert.equal(control.code, 0)
  assert.deepEqual(JSON.parse(control.stdout).findings, [])
})

test('--counting and --duplicate-scope are wired through', async () => {
  const graphemes = await cli(['--config', fixture('counting', 'meta.config.json'), '--json'])
  const codePoints = await cli(['--config', fixture('counting', 'meta.config.json'), '--json', '--counting', 'codePoints'])
  const withinLocale = await cli(['--config', fixture('scope', 'meta.config.json'), '--json'])
  const acrossSite = await cli(['--config', fixture('scope', 'meta.config.json'), '--json', '--duplicate-scope', 'site'])

  assert.equal(graphemes.code, 0)
  assert.deepEqual(rules(JSON.parse(graphemes.stdout)), [])
  assert.equal(codePoints.code, 1)
  assert.deepEqual(rules(JSON.parse(codePoints.stdout)), ['title-too-long'])

  assert.equal(withinLocale.code, 0)
  assert.equal(JSON.parse(withinLocale.stdout).summary.duplicateGroups, 0)
  assert.equal(acrossSite.code, 1)
  assert.equal(JSON.parse(acrossSite.stdout).summary.duplicateGroups, 1)
})

test('--root decides where page paths resolve', async () => {
  const correct = await cli(['--config', example('clean', 'meta.config.json'), '--json', '--root', example('clean')])
  const wrong = await cli(['--config', example('clean', 'meta.config.json'), '--json', '--root', example()])

  assert.equal(correct.code, 0)
  assert.equal(JSON.parse(correct.stdout).summary.checked, 3)
  assert.equal(wrong.code, 2)
  assert.deepEqual(rules(JSON.parse(wrong.stdout)), [
    'nothing-checked',
    'page-unreadable',
    'page-unreadable',
    'page-unreadable',
  ])
})

test('an unsupported counting unit or scope on the command line is a configuration error', async () => {
  const badUnit = await cli(['--config', example('clean', 'meta.config.json'), '--counting', 'bytes'])
  const badScope = await cli(['--config', example('clean', 'meta.config.json'), '--duplicate-scope', 'page'])

  assert.equal(badUnit.code, 2)
  assert.equal(badUnit.stdout, '')
  assert.match(badUnit.stderr, /counting must be one of graphemes, codePoints, utf16CodeUnits/)
  assert.equal(badScope.code, 2)
  assert.equal(badScope.stdout, '')
  assert.match(badScope.stderr, /duplicateScope must be one of locale, site/)
})


test/html.test.mjs
import assert from 'node:assert/strict'
import test from 'node:test'

import { scanHead } from '../src/html.mjs'

const LIMITS = { maxHeadLength: 100000, maxTags: 1000 }

function head(body) {
  return `<!DOCTYPE html>\n<html lang="en">\n<head>\n${body}\n</head>\n<body><p>x</p></body>\n</html>\n`
}

test('the title and description are read with their source positions', () => {
  const result = scanHead(head('<title>Edilec</title>\n<meta name="description" content="A description." />'), LIMITS)

  assert.equal(result.title.text, 'Edilec')
  assert.deepEqual(result.title.position, { line: 4, column: 1 })
  assert.equal(result.description.text, 'A description.')
  assert.deepEqual(result.description.position, { line: 5, column: 1 })
  assert.equal(result.headEnded, true)
  assert.equal(result.truncated, null)
})

test('a title inside a script is not the document title', () => {
  const result = scanHead(
    head('<script>const decoy = "<title>fake</title>";</script>\n<title>Real</title>'),
    LIMITS,
  )

  assert.equal(result.title.text, 'Real')
  assert.equal(result.titleElements, 1)
})

test('a commented-out title is ignored', () => {
  const result = scanHead(head('<!-- <title>commented</title> -->\n<title>Real</title>'), LIMITS)

  assert.equal(result.title.text, 'Real')
  assert.equal(result.titleElements, 1)
})

test('a greater-than inside a quoted attribute does not end the tag', () => {
  const result = scanHead(head('<meta name="description" content="5 > 3 and 2 < 4" />'), LIMITS)

  assert.equal(result.description.text, '5 > 3 and 2 < 4')
})

test('repeated elements are counted and the first one wins', () => {
  const result = scanHead(
    head('<title>First</title>\n<title>Second</title>\n<meta name="description" content="one" />\n<meta name="Description" content="two" />'),
    LIMITS,
  )

  assert.equal(result.title.text, 'First')
  assert.equal(result.titleElements, 2)
  assert.equal(result.description.text, 'one')
  assert.equal(result.descriptionElements, 2)
})

test('a meta description with no content attribute is present but empty', () => {
  const result = scanHead(head('<title>x</title>\n<meta name="description" />'), LIMITS)

  assert.equal(result.descriptionElements, 1)
  assert.equal(result.description.text, null)
})

test('the charset is read from either spelling', () => {
  const direct = scanHead(head('<meta charset="iso-8859-1" /><title>x</title>'), LIMITS)
  const equivalent = scanHead(
    head('<meta http-equiv="Content-Type" content="text/html; charset=windows-1252" /><title>x</title>'),
    LIMITS,
  )

  assert.equal(direct.charset.value, 'iso-8859-1')
  assert.equal(equivalent.charset.value, 'windows-1252')
})

test('scanning stops at the body even when the head was never closed', () => {
  const source = '<html lang="en"><title>x</title><body><h1>y</h1><title>not counted</title></body>'
  const result = scanHead(source, LIMITS)

  assert.equal(result.headEnded, true)
  assert.equal(result.titleElements, 1)
  assert.equal(result.truncated, null)
})

test('the lang attribute is read from the html element with its position', () => {
  const result = scanHead(head('<title>x</title>'), LIMITS)

  assert.equal(result.lang.value, 'en')
  assert.deepEqual(result.lang.position, { line: 2, column: 1 })
})

test('running past maxHeadLength before the head ends is reported, not silently truncated', () => {
  const source = head(`<title>x</title>${'<link rel="preload" href="/a.js" />'.repeat(200)}`)
  const result = scanHead(source, { maxHeadLength: 120, maxTags: 1000 })

  assert.equal(result.truncated, 'maxHeadLength')
  assert.equal(result.headEnded, false)
})

test('running past maxTags before the head ends is reported', () => {
  const source = head(`<title>x</title>${'<link rel="preload" href="/a.js" />'.repeat(200)}`)
  const result = scanHead(source, { maxHeadLength: 100000, maxTags: 5 })

  assert.equal(result.truncated, 'maxTags')
  assert.equal(result.tagsScanned, 6)
})

test('a document shorter than the bound whose head simply never closes is not truncated', () => {
  const result = scanHead('<html lang="en"><head><title>x</title>', { maxHeadLength: 100000, maxTags: 1000 })

  assert.equal(result.truncated, null)
  assert.equal(result.title.text, 'x')
})

test('scanHead refuses limits that are not positive integers', () => {
  assert.throws(() => scanHead('<title>x</title>', { maxHeadLength: 0, maxTags: 1 }), /maxHeadLength/)
  assert.throws(() => scanHead('<title>x</title>', { maxHeadLength: 1, maxTags: 0 }), /maxTags/)
})


src/rules.mjs
/**
 * The rule catalog, the severity table, and the ordering the report depends on.
 *
 * Severity is the whole difference between a run that fails and one that
 * passes. Left as a literal at every construction site it drifts silently: one
 * edit flips a refusal into a green build and no test notices. So there is
 * exactly one table, every finding takes its severity from it, an unknown rule
 * throws instead of defaulting, and `docs/meta-snippet-rules.md` is asserted
 * against this table in both directions.
 */

export const SEVERITIES = Object.freeze(['error', 'warning', 'info'])

/**
 * A length bound is an error because the operator declared it.
 *
 * The bound itself is house style and the report says so in as many words: no
 * search engine publishes a character limit. But a linter that only ever warns
 * about the policy it was handed does not enforce that policy at all, and a
 * configured bound that cannot fail a build is a configured bound nobody reads.
 * The honesty belongs in the wording, not in a severity that makes the check
 * inert.
 */
export const RULE_SEVERITY = Object.freeze({
  'charset-not-utf8': 'warning',
  'description-duplicate': 'error',
  'description-empty': 'error',
  'description-missing': 'error',
  'description-repeated-element': 'warning',
  'description-too-long': 'error',
  'description-too-short': 'error',
  'entity-not-decoded': 'info',
  'head-scan-truncated': 'warning',
  'html-lang-mismatch': 'warning',
  'html-lang-missing': 'info',
  'locale-not-declared': 'error',
  'nothing-checked': 'error',
  'page-limit-exceeded': 'error',
  'page-not-utf8': 'error',
  'page-too-large': 'error',
  'page-unreadable': 'error',
  'text-not-nfc': 'info',
  'text-truncated': 'warning',
  'title-duplicate': 'error',
  'title-empty': 'error',
  'title-missing': 'error',
  'title-repeated-element': 'warning',
  'title-too-long': 'error',
  'title-too-short': 'error',
})

/**
 * Order by UTF-16 code unit, never by locale.
 *
 * `localeCompare` depends on the ICU data compiled into the running Node
 * build, so the same inputs can order differently on two machines. Code-unit
 * order is a property of the strings themselves.
 */
export function byCodeUnit(left, right) {
  if (left === right) return 0
  return left < right ? -1 : 1
}

export const RULE_IDS = Object.freeze([...Object.keys(RULE_SEVERITY)].sort(byCodeUnit))

export const EVIDENCE_LIMIT = 160

/** How much input an excerpt will ever scan, however long the value is. */
export const SCRUB_LIMIT = 8192

/**
 * Characters that would break a line, a terminal or a log reader: the C0 and
 * C1 control ranges plus the two Unicode line separators. Written as code
 * point arithmetic rather than a character class so the source file itself can
 * never contain the control characters it is describing.
 */
function isUnprintable(code) {
  return code < 0x20 || (code >= 0x7f && code <= 0x9f) || code === 0x2028 || code === 0x2029
}

export function severityOf(ruleId) {
  if (!Object.hasOwn(RULE_SEVERITY, ruleId)) {
    throw new RangeError(`Unknown ruleId "${ruleId}"`)
  }
  return RULE_SEVERITY[ruleId]
}

/**
 * A bounded, single-line excerpt.
 *
 * Page content is data. It is never echoed at length, and never echoed in a
 * shape that could be mistaken for an instruction to whoever reads the report.
 */
export function excerpt(value) {
  // Bounded work as well as bounded output: the caller may hand this an
  // attribute value straight out of a two-megabyte document.
  const source = String(value).slice(0, SCRUB_LIMIT)
  let scrubbed = ''
  for (let index = 0; index < source.length; index += 1) {
    scrubbed += isUnprintable(source.charCodeAt(index)) ? ' ' : source[index]
  }
  const flattened = scrubbed.replace(/\s+/g, ' ').trim()
  if (flattened.length <= EVIDENCE_LIMIT) return flattened
  return `${flattened.slice(0, EVIDENCE_LIMIT)}...`
}

/**
 * Build a location. `line` and `column` are 1-based and tool-specific: the
 * report contract defines `file` and `pointer`, and this tool adds the source
 * position of the element a finding is about, because "which tag" is the first
 * thing a reader needs.
 */
export function at(file, pointer, position = null) {
  const location = {}
  if (file !== null && file !== undefined) location.file = file
  if (pointer !== null && pointer !== undefined) location.pointer = pointer
  if (position !== null && position !== undefined) {
    location.line = position.line
    location.column = position.column
  }
  return location
}

export function makeFinding(ruleId, message, location, extra = {}) {
  const finding = { ruleId, severity: severityOf(ruleId), message, location }
  if (extra.evidence !== undefined && extra.evidence !== null) finding.evidence = excerpt(extra.evidence)
  if (extra.suggestion !== undefined && extra.suggestion !== null) finding.suggestion = extra.suggestion
  return finding
}

/**
 * The documented sort key: file, pointer, ruleId, line, column, message.
 *
 * A missing file or pointer sorts as the empty string and a missing position
 * sorts as 0, so config-level findings come first and every run over the same
 * inputs serializes to the same bytes.
 */
export function compareFindings(left, right) {
  const keys = [
    byCodeUnit(left.location?.file ?? '', right.location?.file ?? ''),
    byCodeUnit(left.location?.pointer ?? '', right.location?.pointer ?? ''),
    byCodeUnit(left.ruleId, right.ruleId),
    (left.location?.line ?? 0) - (right.location?.line ?? 0),
    (left.location?.column ?? 0) - (right.location?.column ?? 0),
    byCodeUnit(left.message, right.message),
  ]
  for (const key of keys) {
    if (key !== 0) return key
  }
  return 0
}
